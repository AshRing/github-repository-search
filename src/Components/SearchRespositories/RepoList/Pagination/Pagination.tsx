import * as React from "react";

import { ArrowContainer, PaginationContainer } from "./Pagination.styles";
import {
	faAngleDoubleLeft,
	faAngleDoubleRight,
	faAngleLeft,
	faAngleRight,
} from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

interface Props {
	changePage: (direction: number) => void;
	currentPage: number;
	totalPages: number;
}

export const Pagination = ({ changePage, currentPage, totalPages }: Props) => {
	return (
		<PaginationContainer>
			<ArrowContainer>
				{currentPage > 1 && (
					<>
						<button id="pageDoubleLeftArrow" onClick={() => changePage(1)}>
							<FontAwesomeIcon icon={faAngleDoubleLeft} />
						</button>
						<button id="pageLeftArrow" onClick={() => changePage(currentPage - 1)}>
							<FontAwesomeIcon icon={faAngleLeft} />
						</button>
					</>
				)}
			</ArrowContainer>
			<div>
				{currentPage} of {totalPages}
			</div>
			<ArrowContainer>
				{currentPage < totalPages && (
					<>
						<button id="pageRightArrow" onClick={() => changePage(currentPage + 1)}>
							<FontAwesomeIcon icon={faAngleRight} />
						</button>
						<button id="pageDoubleRightArrow" onClick={() => changePage(totalPages)}>
							<FontAwesomeIcon icon={faAngleDoubleRight} />
						</button>
					</>
				)}
			</ArrowContainer>
		</PaginationContainer>
	);
};
