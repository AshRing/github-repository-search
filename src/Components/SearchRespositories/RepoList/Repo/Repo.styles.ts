import { mediaQueries } from "../../../../_style";
import { rgba } from "polished";
import styled from "styled-components";

export const RepoContainer = styled.div`
	padding: 0.5rem 1rem;
	width: 100%;
	background: #fff;
	box-shadow: 2px 2px 2px ${({ theme }) => rgba(theme.colors.primary, 0.3)};
	border-radius: 0.25rem;
	margin-bottom: 1rem;
	overflow: hidden;

	${mediaQueries.desktop} {
		padding: 1rem 1.5rem;
		margin-bottom: 1.5rem;
	}
`;

export const RepoFlexContainer = styled.div`
	display: flex;
	align-items: center;
	justify-content: space-between;
`;

export const RepoName = styled.h2`
	color: ${({ theme }) => theme.colors.primary};
	word-break: break-word;
`;

export const RepoLanguage = styled.span`
	font-style: italic;
`;

export const RepoDescription = styled.p`
	font-size: 0.75rem;
	margin: 0.5rem 0;
`;

export const RepoStars = styled.div`
	display: flex;
	align-items: center;

	svg {
		margin-left: 0.25rem;
		color: ${({ theme }) => theme.colors.primary};
	}
`;
