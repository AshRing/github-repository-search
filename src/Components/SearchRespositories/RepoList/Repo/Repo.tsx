import * as React from "react";

import {
	RepoContainer,
	RepoDescription,
	RepoFlexContainer,
	RepoLanguage,
	RepoName,
	RepoStars,
} from "./Repo.styles";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IRepository } from "../../../../_types";
import { faStar } from "@fortawesome/free-solid-svg-icons";

interface Props {
	repoInfo: IRepository;
}

export const Repo = ({ repoInfo }: Props) => {
	return (
		<RepoContainer>
			<RepoFlexContainer>
				<RepoName>{repoInfo.full_name}</RepoName>
			</RepoFlexContainer>
			<RepoDescription>{repoInfo.description}</RepoDescription>
			<RepoFlexContainer>
				<RepoLanguage>{repoInfo.language}</RepoLanguage>
				<RepoStars>
					<span>{repoInfo.stargazers_count}</span>
					<FontAwesomeIcon icon={faStar} />
				</RepoStars>
			</RepoFlexContainer>
		</RepoContainer>
	);
};
