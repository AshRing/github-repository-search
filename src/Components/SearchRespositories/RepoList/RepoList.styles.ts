import styled from "styled-components";

export const RepoListContainer = styled.div`
	padding-top: 1rem;
	border-top: 1px solid ${({ theme }) => theme.colors.secondary};
`;
