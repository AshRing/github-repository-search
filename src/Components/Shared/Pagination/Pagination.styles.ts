import styled from "styled-components";

export const PaginationContainer = styled.div`
	display: flex;
	justify-content: space-between;
	align-items: center;
	margin: 1rem 0;
`;

export const ArrowContainer = styled.div`
	width: 3.5rem;
	display: flex;

	button {
		margin: 0.25rem;
		background: none;
	}

	svg {
		color: ${({ theme }) => theme.colors.primary};
		font-size: 2rem;
	}
`;
