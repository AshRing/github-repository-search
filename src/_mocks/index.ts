export * from "./repo";
export * from "./TestWrapper";
export * from "./createDocumentListenersMock";
