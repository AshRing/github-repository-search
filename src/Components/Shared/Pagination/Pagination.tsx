import * as React from "react";

import { ArrowContainer, PaginationContainer } from "./Pagination.styles";
import {
	faAngleDoubleLeft,
	faAngleDoubleRight,
	faAngleLeft,
	faAngleRight,
} from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

interface Props {
	changePage: (direction: number) => void;
	currentPage: number;
	totalPages: number;
}

export const Pagination = ({ changePage, currentPage, totalPages }: Props) => {
	return (
		<PaginationContainer>
			<ArrowContainer>
				{currentPage > 1 && (
					<>
						<button onClick={() => changePage(1)}>
							<FontAwesomeIcon icon={faAngleDoubleLeft} />
						</button>
						<button onClick={() => changePage(currentPage - 1)}>
							<FontAwesomeIcon icon={faAngleLeft} />
						</button>
					</>
				)}
			</ArrowContainer>
			<div>
				{currentPage} of {totalPages}
			</div>
			<ArrowContainer>
				{currentPage < totalPages && (
					<>
						<button onClick={() => changePage(currentPage + 1)}>
							<FontAwesomeIcon icon={faAngleRight} />
						</button>
						<button onClick={() => changePage(totalPages)}>
							<FontAwesomeIcon icon={faAngleDoubleRight} />
						</button>
					</>
				)}
			</ArrowContainer>
		</PaginationContainer>
	);
};
