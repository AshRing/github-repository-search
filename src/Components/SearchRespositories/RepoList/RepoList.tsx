import * as React from "react";

import { RepoListContainer, StatusText } from "./RepoList.styles";

import { IRepository } from "../../../_types";
import { Link } from "react-router-dom";
import { Pagination } from "./Pagination";
import { Repo } from "./Repo";

interface Props {
	changePage: (pageNum: number) => void;
	pageNum: number;
	repos: IRepository[];
	reposLoading: boolean;
	totalPages: number;
}

export const RepoList = ({ changePage, pageNum, repos, reposLoading, totalPages }: Props) => {
	const renderStatusText = () => {
		if (reposLoading) {
			return <StatusText>Loading...</StatusText>;
		}

		if (!repos) {
			return <StatusText>Please enter a search term</StatusText>;
		}
	};

	return (
		<RepoListContainer>
			{renderStatusText()}
			{!reposLoading && (
				<>
					{repos?.map((repo: IRepository) => (
						<Link to={`/${repo.owner.login}/${repo.name}`} key={repo.id}>
							<Repo repoInfo={repo} />
						</Link>
					))}
					{repos?.length > 0 && (
						<Pagination
							changePage={changePage}
							currentPage={pageNum}
							totalPages={totalPages}
						/>
					)}
				</>
			)}
		</RepoListContainer>
	);
};
