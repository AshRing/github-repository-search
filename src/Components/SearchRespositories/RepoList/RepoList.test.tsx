import * as React from "react";

import { TestWrapper, mockRepo1 } from "../../../_mocks";
import { mount, shallow } from "enzyme";

import { Pagination } from "./Pagination";
import { Repo } from "./Repo";
import { RepoList } from ".";
import { StatusText } from "./RepoList.styles";

let props;
beforeEach(() => {
	props = {
		changePage: jest.fn(),
		pageNum: 1,
		repos: [mockRepo1],
		reposLoading: false,
		totalPages: 1,
	};
});

test("RepoList matches snapshot", () => {
	const wrapper = shallow(<RepoList {...props} />);
	expect(wrapper).toMatchSnapshot();
});

test("if repos are loading, display correct text & components", () => {
	props = {
		...props,
		reposLoading: true,
	};
	const wrapper = mount(
		<TestWrapper>
			<RepoList {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find(StatusText).text()).toBe("Loading...");
	expect(wrapper.find(Repo).exists()).toBeFalsy();
	expect(wrapper.find(Pagination).exists()).toBeFalsy();
});

test("if repos are not loading, and are undefined, display correct text & components", () => {
	props = {
		...props,
		reposLoading: false,
		repos: undefined,
	};
	const wrapper = mount(
		<TestWrapper>
			<RepoList {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find(StatusText).text()).toBe("Please enter a search term");
	expect(wrapper.find(Repo).exists()).toBeFalsy();
	expect(wrapper.find(Pagination).exists()).toBeFalsy();
});

test("if repos are not loading and are NOT undefined, display correct components", () => {
	props = {
		...props,
		reposLoading: false,
		repos: [mockRepo1],
	};
	const wrapper = mount(
		<TestWrapper>
			<RepoList {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find(StatusText).exists()).toBeFalsy();
	expect(wrapper.find(Repo).exists()).toBeTruthy();
	expect(wrapper.find(Repo)).toHaveLength(props.repos.length);
	expect(wrapper.find(Pagination).exists()).toBeTruthy();
});
