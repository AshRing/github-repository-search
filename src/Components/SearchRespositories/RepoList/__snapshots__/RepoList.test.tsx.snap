// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoList matches snapshot 1`] = `ShallowWrapper {}`;
